src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs

```diff
@@ -279,9 +282,72 @@ private void HandleInitialFixupRequest(string id, string paramsJson)
             }));
         }
 
+        private void HandleCodeActionRequest(string id, string paramsJson)
+        {
+            if (id == null)
+            {
+                _sendToClient(JsonRpcHelper.CreateErrorResult(id, JsonRpcHelper.ErrorCode.InvalidRequest));
+                return;
+            }
+
+            Contracts.AssertValue(id);
+            Contracts.AssertValue(paramsJson);
+
+            if (!TryParseParams(paramsJson, out CodeActionParams codeActionParams))
+            {
+                _sendToClient(JsonRpcHelper.CreateErrorResult(id, JsonRpcHelper.ErrorCode.ParseError));
+                return;
+            }
+
+            var documentUri = codeActionParams.TextDocument.Uri;
+            var scope = _scopeFactory.GetOrCreateInstance(documentUri);
+
+            var uri = new Uri(documentUri);
+            var expression = HttpUtility.ParseQueryString(uri.Query).Get("expression");
+            if (expression == null)
+            {
+                _sendToClient(JsonRpcHelper.CreateErrorResult(id, JsonRpcHelper.ErrorCode.InvalidParams));
+                return;
+            }
+
+            var codeActions = new Dictionary<string, CodeAction[]>();
+            foreach (var codeActionKind in codeActionParams.Context.Only)
+            {
+                switch (codeActionKind)
+                {
+                    case CodeActionKind.QuickFix:
+                        var result = scope.QuickFix(expression);
+
+                        var items = new List<CodeAction>();
+
+                        foreach (var item in result)
+                        {
+                            items.Add(new CodeAction()
+                            {
+                                Title = item.Title,
+                                Kind = codeActionKind,
+                                Edit = new WorkspaceEdit
+                                {
+                                    Changes = new Dictionary<string, TextEdit[]> { { documentUri, new[] { new TextEdit { Range = codeActionParams.Range, NewText = item.Text } } } }
```

Comment:
> Range = codeActionParams.Range, [](http://example.com/codeflow?start=120&length=32)

note: this is where we need to update to have more specific ranges on the code fix.

Reply:
Now, I have updated object model to return the location of fix.